SPECS/shim/TESTING

```diff
@@ -10,21 +10,216 @@ separate keys, and the public part of the key used to sign mm/fb
-(public) key. However, since signing mm/fb with the Azure Linux
-signing key makes the mm/fb binaries usable on actual production
-systems, so using the signing key for development work is *highly*
-discouraged. Generally this should only be done with an updated shim
-package after all other testing is complete, *only* to validate that
-the shim will still correctly load the mm/fb binaries using the Azure
-Linux signing key.
-
-The second option, which should be used during normal shim
-development, is to generate a local signing key and embed the public
-portion inside the shim.
-
-1) 
+Note that these instructions are for x86-64, for aarch64 systems,
+replace all 'x64' with 'aa64'. Also, replace all instances of $VERSION
+with the appropriate version number.
+
+
+
+1) Generate local signing certificates
+
+Install required packages:
+
+$ sudo dnf install pesign nss-tools efivar
+
+Create local signing certificates for testing, using the 'make-certs'
+script from the pesign source (replace '/usr/src/azl' with whatever
+the rpm %_topdir macro location is for your system):
+
+$ sudo dnf download --source pesign
+$ sudo rpm -i pesign-$VERSION
+$ cd /usr/src/azl
+$ sudo rpmbuild -bp SPECS/pesign.spec
+$ cd BUILD/pesign-$VERSION
+$ cd src/certs
+$ sudo ./make-certs KeyInDB azurelinux@microsoft.com all codesign 1.3.6.1.4.1.311.10.3.1
+$ sudo ./make-certs KeyInShim azurelinux@microsoft.com all codesign 1.3.6.1.4.1.311.10.3.1
+
+Now, optionally, remove your pesign DB (you only need to do this if
+you might have old stuff in it):
+
+$ sudo rm -rfv /etc/pki/pesign/*
+
+Now put the root CA and the signing certificates into your pesign db:
+
+$ sudo certutil -d /etc/pki/pesign -A -n 'my CA' -t CT,CT,CT -i ca.crt
+$ sudo pk12util -d /etc/pki/pesign -i KeyInDB.p12
+$ sudo pk12util -d /etc/pki/pesign -i KeyInShim.p12
+$ sudo certutil -d /etc/pki/pesign -A -i KeyInDB.crt -n KeyInDB -t u
+$ sudo certutil -d /etc/pki/pesign -A -i KeyInShim.crt -n KeyInShim -t u
+
+
+
+2) Rebuild the shim-unsigned package
+
+Note that including the KeyInShim inside the shim EFI binary is
+optional; if you do not need to test the shim loading the mm/fb
+binaries, there is no need to sign the mm/fb binaries (since you are
+not testing them) and you do not need to resign the boot loader or
+kernel (since you are building the shim with the Azure Linux
+production certificate embedded), and you can proceed to the next step.
+
+First, get the KeyInShim certificate in DER format:
+
+$ sudo certutil -d /etc/pki/pesign -L -n KeyInShim -r > key-in-shim.der
+
+Because the shim package can only have a single embedded certificate,
+you have 2 options on how to embed the test certificate in the shim.
+
+Option a) Replace the production Azure Linux certificate
+
+This will result in a shim that contains *only* the test certificate,
+and will only be able to load other EFI binaries signed by the test
+certificate. This means the shim will not be load the
```

Comment:
minor typo, "will not be load" should be "will not load"

Reply:
ack thnx